### PR TITLE
Modify organisation check to support GH Action runner for an enterprise

### PR DESCRIPTION
**Pull request by hamishr**
## Description

I've changed the organisation conditional check from a boolean to match a string

## Type of change
- [x] New feature (non-breaking change which adds functionality)

This requires the companion variables to be defined for the enterprise eg.

```
---
github_url: "https://github.example.com"
github_api_url: "https://github.example.com/api/v3"
runner_type: "ent"
github_ent: "enterprise_name"
```

Refs:
 * https://gh.sdintra.net/SysAdmin/DevOps/issues/448

### CODE REVIEW

File: defaults/main.yml [hunk 1]
```diff
@@ -29,8 +29,8 @@ github_api_url: "https://api.github.com"
 # Personal Access Token for your GitHub account
 access_token: "{{ lookup('env', 'PERSONAL_ACCESS_TOKEN') }}"
 
-# Is it the runner for organization or not?
-runner_org: no
+# Is it the runner for a "repo", an "org" organization or an "ent" enterprise?
```
Comment (adamc00) [inline defaults/main.yml]: Let's use the full names, `repository`, `organisation`, `enterprise`. There's no benefit to these abbreviations and they they add the to opportunity for misunderstanding.

File: tasks/collect_info_ent.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+  - name: Check service facts
+    service_facts:
+    tags:
+      - install
+      - uninstall
+
+  - name: Build service name
+    set_fact:
+      runner_service: "actions.runner.{{ ([( github_ent | default(github_account))[:45], runner_name] | join('.'))[:57]  }}.service"
```
Comment (adamc00) [inline tasks/collect_info_ent.yml]: Could `github_account` just be set to the enterprise name in the `ent` context? 

If we'd never need to set both `github_ent` and `github_account` then we are adding to the potential for misconfiguration / misunderstanding for no reason. If there is a need to set both then that's fine. When is it required?

**Review by adamc00 — CHANGES_REQUESTED** [review]

File: defaults/main.yml (hunk 1 above)
Comment (hamishr) [reply defaults/main.yml]: Fair point, I've fixed this up in the latest commit.

File: tasks/collect_info_ent.yml (hunk 2 above)
Comment (hamishr) [reply tasks/collect_info_ent.yml]: Yes, no need for github_account when we are using the enterprise token.
I've removed this, and removed the assertion to supply the github_account at all when using an enterprise runner.

File: tasks/assert.yml [hunk 3]
```diff
@@ -2,11 +2,21 @@
 - name: Check variables (RUN ONCE)
   assert:
     that:
-      - github_account is defined
```
Comment (adamc00) [inline tasks/assert.yml]: Why can't we just allow `github_account` to be set to the enterprise name in the enterprise case. That's what they do for the repository and organisation case as far as I can tell. Why do we need  to introduce the complexity of another variable? That's what my previous feedback was trying to point out.

File: defaults/main.yml [hunk 4]
```diff
@@ -29,8 +29,8 @@ github_api_url: "https://api.github.com"
 # Personal Access Token for your GitHub account
 access_token: "{{ lookup('env', 'PERSONAL_ACCESS_TOKEN') }}"
 
-# Is it the runner for organization or not?
-runner_org: no
+# Is it the runner for a "repository", an "organization" or an "enterprise"?
+runner_type: "repository"
```
Comment (martin-andrew-ellis) [inline defaults/main.yml]: The problem with out right removal of `runner_org` is that you will require all existing users that have set it change their code.

I'm not sure what @adamc00 thinks but you may want something like:
```
# Is it the runner for a "repository", an "organization" or an "enterprise"?
runner_type: "{{ (runner_org | default('no')) ? "organization" : "repository }}"
```

File: tasks/install_runner.yml [hunk 5]
```diff
@@ -86,7 +86,9 @@
   become: yes
   become_user: "{{ runner_user }}"
   no_log: "{{ hide_sensitive_logs | bool }}"
-  when: runner_name not in registered_runners.json.runners|map(attribute='name')|list and not runner_org
+  when:
+   - runner_name not in registered_runners.json.runners|map(attribute='name')|list
```
Comment (martin-andrew-ellis) [inline tasks/install_runner.yml:90]: I like the fact that you broken down the explicit `and`s into lists, but I feel there is a lot repeated code here and its doing my head in to look for the `in` and `not in`.

Could we:
* contain the logic in some facts
* use block to group similarly conditioned tasks

```
  set_fact:
     is_registered_runner: "{{ runner_name n registered_runners.json.runners|map(attribute='name')|list }}"

...

- name: Register a new service
  when: not is_registered_runner
  block:
    - name: Register runner (if new installation) for repository
       ...
      when: runner_type == "repository"

    ...

- name: Replace registered runner
  when:
    - is_registered_runner
    - reinstall_runner
      
  ...
```

**Review by martin-andrew-ellis — CHANGES_REQUESTED** [review]

File: tasks/assert.yml (hunk 3 above)
Comment (martin-andrew-ellis) [reply tasks/assert.yml]: I'm not sure what you mean here @adamc00 is allowed just not required when its for the enterprise. 

Or is this to do with fetching the token?'

Comment (adamc00) [reply tasks/assert.yml]: Why create an additional variable when they are already using this one for the account or organisation name depending on the type of runner they are installing? We are just adding complexity by doing so as far as I can tell. Why not just allow the value of `github_account` to be the enterprise name when installing for an enterprise, following on from the pattern they already use.

This is based on the assumption that we don't need both values for some reason but we don't appear to.

File: defaults/main.yml (hunk 4 above)
Comment (hamishr) [reply defaults/main.yml]: Thanks @martin-andrew-ellis 
I've applied the changes discussed, following the example you provided.
The bulk of which is in `tasks/asserts.yml`

File: tasks/assert.yml (hunk 3 above)
Comment (hamishr) [reply tasks/assert.yml]: Yup, correct @adamc00  I misunderstood this.
Have fixed it up in the current commit, and tested.

File: tasks/install_runner.yml (hunk 5 above)
Comment (hamishr) [reply tasks/install_runner.yml:90]: Here I was mirroring the practices in the upstream repo.
It's annoying to read when the tasks aren't blocked together, but I think we're better to leave it that way.

Comment (martin-andrew-ellis) [reply tasks/install_runner.yml:90]: From Stephen Thirlwwall:
> If I get a drive-by PR with some useful feature and they've taken it upon themselves to reorganise/refactor large chunks of the code, I'm far less likely to accept that PR.
> I'd be inclined to make it a two parter. First implement your feature using their style. Then if that gets accepted, make another PR with the whole lot refactored into blocks.

I think what he has said makes sense so, list leave it as is, For this PR.

File: tasks/collect_info_ent.yml [hunk 6]
```diff
@@ -0,0 +1,55 @@
+---
+- block:
+  - name: Get registration token (RUN ONCE)
+    uri:
+      url: "{{ github_api_url }}/enterprises/{{ github_owner | default(github_account) }}/actions/runners/registration-token"
```
Comment (martin-andrew-ellis) [inline tasks/collect_info_ent.yml:5]: Every time I see `github_owner`, it is in the context of `github_owner | default(github_account)`

Given that would it be better to set in `defaults/main.yml`
```
github_owner: "{{ github_account }}"
```
?

Comment (martin-andrew-ellis) [reply tasks/collect_info_ent.yml:5]: Actually I'm going withdraw this objection.  I can see we are only being consistent with the existing code base.

Comment (martin-andrew-ellis) [reply tasks/collect_info_ent.yml:5]: Might be a refactor we can offer at a later date.

**Review by martin-andrew-ellis — APPROVED** [review]
I think this covers all I've talked about.

File: tasks/collect_info_ent.yml (hunk 6 above)
Comment (hamishr) [reply tasks/collect_info_ent.yml:5]: That, and I think we can remove the assertion that the github_account is defined if the github_owner is already defined.
(Not forgetting your suggestion of splitting tasks into blocks more clearly)

**Pull request merged**